pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,213 @@
+// all endpoints which are part of the same cluster.
+type IPCache struct {
+	mutex              lock.RWMutex
+	ipToIdentityCache  map[string]identity.NumericIdentity
+	identityToIPCache  map[identity.NumericIdentity]map[string]struct{}
+	xdsResourceMutator xds.ResourceMutator
+}
+
+// NewIPCache returns a new IPCache with the mappings of endpoint IP to security
+// identity (and vice-versa) initialized, along with the provided resourceMutator.
+func NewIPCache(resourceMutator xds.ResourceMutator) *IPCache {
+	return &IPCache{
+		ipToIdentityCache:  map[string]identity.NumericIdentity{},
+		identityToIPCache:  map[identity.NumericIdentity]map[string]struct{}{},
+		xdsResourceMutator: resourceMutator,
+	}
+}
+
+// Upsert adds / updates  the provided IP and identity into both caches contained
+// within ipc.
+func (ipc *IPCache) Upsert(endpointIP string, identity identity.NumericIdentity) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+
+	// Update both maps.
+	ipc.ipToIdentityCache[endpointIP] = identity
+
+	_, found := ipc.identityToIPCache[identity]
+	if !found {
+		ipc.identityToIPCache[identity] = map[string]struct{}{}
+	}
+	ipc.identityToIPCache[identity][endpointIP] = struct{}{}
+	endpointIPs := ipc.identityToIPCache[identity]
+
+	// Sort IPs for Upsert into xdsResourceMutator.
+	ipStrings := make([]string, 0, len(endpointIPs))
+	for endpointIP := range endpointIPs {
+		ipStrings = append(ipStrings, endpointIP)
+	}
+	sort.Strings(ipStrings)
+	ipc.xdsResourceMutator.Upsert(envoy.NetworkPolicyHostsTypeURL, identity.StringID(), &envoyAPI.NetworkPolicyHosts{Policy: uint64(identity), HostAddresses: ipStrings}, false)
+
+}
+
+// Delete removes the provided IP-to-security-identity mapping from both caches
+// within ipc.
+func (ipc *IPCache) Delete(endpointIP string) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+
+	identity, found := ipc.ipToIdentityCache[endpointIP]
+	if found {
+		delete(ipc.ipToIdentityCache, endpointIP)
+		delete(ipc.identityToIPCache[identity], endpointIP)
+		if len(ipc.identityToIPCache[identity]) == 0 {
+			delete(ipc.identityToIPCache, identity)
+		}
+		ipc.xdsResourceMutator.Delete(envoy.NetworkPolicyHostsTypeURL, identity.StringID(), false)
+	}
+}
+
+// LookupByIP returns the corresponding security identity that endpoint IP maps
+// to within the provided IPCache, as well as if the corresponding entry exists
+// in the IPCache.
+func (ipc *IPCache) LookupByIP(endpointIP string) (identity.NumericIdentity, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	identity, exists := ipc.ipToIdentityCache[endpointIP]
+	return identity, exists
+}
+
+// LookupByIdentity returns the set of endpoint IPs that have security identity
+// ID, as well as if the corresponding entry exists in the IPCache.
+func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (map[string]struct{}, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	ips, exists := ipc.identityToIPCache[id]
```

Comment:
Same here

Reply:
I tried this initially, but can't because map lookup has a variable number of values returned.